src/EliminateBoolVectors.cpp

```diff
@@ -0,0 +1,146 @@
+#include "IRMutator.h"
+#include "IROperator.h"
+
+namespace Halide {
+namespace Internal {
+
+class EliminateBoolVectors : public IRMutator {
+private:
+    using IRMutator::visit;
+
+    template <typename T>
+    void visit_comparison(const T* op) {
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+        Type t = a.type();
+
+        // Ensure both a and b have the same type (if this is a vector
+        // comparison). This should only be necessary if the operands are
+        // integer vectors (promoted from bool vectors).
+        if (t.lanes() > 1 && t.bits() != b.type().bits()) {
+            internal_assert(t.is_int() && b.type().is_int());
+
+            t = t.with_bits(std::max(t.bits(), b.type().bits()));
+            if (t != a.type()) {
+                a = Cast::make(t, a);
+            }
+            if (t != b.type()) {
+                b = Cast::make(t, b);
+            }
+        }
+
+        if (!a.same_as(op->a) || !b.same_as(op->b)) {
+            expr = T::make(a, b);
+        } else {
+            expr = op;
+        }
+
+        if (t.lanes() > 1) {
+            // To represent bool vectors, OpenCL uses vectors of signed
+            // integers with the same width as the types being compared.
+            t = t.with_code(Type::Int);
+            expr = Cast::make(t, expr);
```

Comment:
Yes, because lets's say we have a compare of the type   (v32xi16 > v32i16). This generate a result of v32i1 i.e a boolean vector. We sign extend it into v32i16 using this pass. i.e. v32i16(v32i1). So we have a compare node i.e cast(v32xi16, (v32xi16 > v32i16)). We lower this condition using an HVX intrinsic like vcmp.gt by first bitcasting the v32i16 into v512i1 and then generating a vselect.